File: docs/palindromes-in-peg/scaffold_window_pal.py
"""Two dyadic PAL stages in one transition per original binary input.

GS service is a fixed finite Boolean circuit, with block-window cursors and
counter origins. Historical clocks allocate one cell per actual character;
result packets store the bounded flag emissions of that same transition.
"""
from copy import copy

from gs_batch_clock import DEFAULT_BATCH
from scaffold_circuit import (Circuit, Value, Ref, PREVIOUS, EMPTY, TRUE, FALSE,
                              conjunction as AND, disjunction as OR, neg, choose)
from scaffold_circuit_structs import Stack, StackPool
from scaffold_flag_packets import FlagStack
from scaffold_window_workers import matcher, flags


class WindowStage:
  def __init__(self, circuit, index, history_pool, worker):
    self.circuit, self.index, self.worker = circuit, index, worker
    name = f"stage{index}"
    self.half = Stack(history_pool, name + ".half", "cells")
    self.clock = Stack(history_pool, name + ".clock", "cells")
    self.alive_key, self.interval_key = name + ".alive", name + ".interval"
    self.alive = circuit.get(PREVIOUS, self.alive_key, (False, True), False).eq(True)
    self.interval = circuit.get(PREVIOUS, self.interval_key, tuple(range(7)), 0)
    self.pending_key, self.batch_key = name + ".pending", name + ".batch"
    self.pending = circuit.get(PREVIOUS, self.pending_key, (False, True), False).eq(True)
    self.batch = circuit.get(PREVIOUS, self.batch_key, (0, 1, 2, 3), 0)
    self.results = [FlagStack(worker.flag_pool, name + f".result{i}") for i in range(4)]
    self.birth, self.release, self.middle = FALSE, FALSE, FALSE

  def answering(self):
    return AND(self.alive, OR(*(self.interval.eq(i) for i in range(2, 6))))

  def advance(self, birth, half):
    self.clock.drop(self.alive)
    boundary = AND(self.alive, self.clock.empty())
    self.interval = Value.select(boundary, self.interval.map(lambda n: min(n + 1, 6)), self.interval)
    self.clock.copy_from(self.half, boundary)
    release = AND(boundary, OR(*(self.interval.eq(i) for i in range(1, 5))))
    self.circuit.require(neg(self.pending), release)
    self.batch = Value.select(release, self.interval.map(lambda n: max(0, min(3, n - 1))).recode((0, 1, 2, 3)), self.batch)
    self.alive = AND(self.alive, neg(AND(boundary, self.interval.eq(6))))
    self.half.copy_from(half, birth)
    self.clock.copy_from(half, birth)
    self.interval = Value.select(birth, Value.constant(0), self.interval)
    self.alive = OR(self.alive, birth)
    self.birth, self.release = birth, AND(release, neg(birth))
    self.pending = AND(self.pending, neg(birth))
    for result in self.results: result.clear(birth)

  def consume(self):
    active = self.answering()
    selected = copy(self.results[0])
    selected.root, selected.index = EMPTY, self.worker.flag_pool.index(-1)
    for index, result in enumerate(self.results):
      selected.copy_from(result, self.interval.eq(index + 2))
    self.middle = AND(active, selected.pop(active))
    for index, result in enumerate(self.results):
      result.copy_from(selected, AND(active, self.interval.eq(index + 2)))

  def capture(self):
    self.pending = OR(self.pending, self.release)
    done = AND(self.pending, self.worker.mode.eq("done"))
    for index, result in enumerate(self.results):
      result.copy_from(self.worker.flags, AND(done, self.batch.eq(index)))
    self.pending = AND(self.pending, neg(done))

  def finalize(self):
    self.half.finalize()
    self.clock.finalize()
    for result in self.results: result.finalize()
    for key, value in ((self.interval_key, self.interval), (self.batch_key, self.batch)):
      self.circuit.put(key, value)
    for key, value in ((self.alive_key, self.alive), (self.pending_key, self.pending)):
      self.circuit.put(key, Value.select(value, Value.constant(True), Value.constant(False)))


class WindowPAL:
  def __init__(self, circuit, rates=DEFAULT_BATCH):
    if rates.k != DEFAULT_BATCH.k or rates.matching < DEFAULT_BATCH.matching or rates.flags < DEFAULT_BATCH.flags:
      raise ValueError("the complete PAL source requires the derived k=8 service bounds")
    self.circuit, self.rates = circuit, rates
    self.matchers = [matcher(circuit, f"match{i}", rates.matching) for i in range(2)]
    self.flags = [flags(circuit, f"flag{i}", rates.flags) for i in range(2)]
    history_pool = StackPool(circuit, {"cells": 1})
    self.history = Stack(history_pool, "pal.history", "cells")
    self.power = Stack(history_pool, "pal.power", "cells")
    self.next_birth = Stack(history_pool, "pal.next_birth", "cells")
    self.stages = [WindowStage(circuit, i, history_pool, self.flags[i]) for i in range(2)]
    self.power_ready = circuit.get(PREVIOUS, "pal.power_ready", (False, True), False).eq(True)
    self.slot = circuit.get(PREVIOUS, "pal.slot", (0, 1), 0)
    self.small = circuit.get(PREVIOUS, "pal.small", (0, 1, 2, 3, 4), 0)
    self.first = circuit.get(PREVIOUS, "pal.first", (False, True), False)
    self.output = TRUE

  def tick(self):
    c, current = self.circuit, self.circuit.input().eq("b")
    self.first = Value.select(self.small.eq(0), Value.select(current, Value.constant(True), Value.constant(False)), self.first)
    self.small = self.small.map(lambda n: min(4, n + 1))
    self.history.push()
    first = neg(self.power_ready)
    self.next_birth.drop(self.power_ready)
    birth = AND(self.power_ready, self.next_birth.empty())
    for index, stage in enumerate(self.stages):
      stage.advance(AND(birth, self.slot.eq(index)), self.power)
    self.power.copy_from(self.history, OR(first, birth))
    self.next_birth.copy_from(self.history, OR(first, birth))
    self.power_ready = TRUE
    self.slot = Value.select(birth, self.slot.cycle(), self.slot)

    for stage, matching, flagging in zip(self.stages, self.matchers, self.flags):
      matching.arrive()
      flagging.arrive()
      flagging.reset_flags(stage.birth)
      matching.start(stage.birth)
      flagging.start(stage.release)
      flagging.mark(stage.release)
      stage.consume()
      matching.service()
      flagging.service()
      stage.capture()

    active = [stage.answering() for stage in self.stages]
    c.require(AND(OR(*active), neg(AND(*active))), self.small.eq(4))
    ordinary = OR(*(AND(enabled, stage.middle, matching.output)
                    for enabled, stage, matching in zip(active, self.stages, self.matchers)))
    small = OR(self.small.eq(1), AND(OR(self.small.eq(2), self.small.eq(3)),
                                  choose(current, self.first.eq(True), self.first.eq(False))))
    self.output = choose(self.small.eq(4), ordinary, small)

  def finalize(self):
    for matching, flagging, stage in zip(self.matchers, self.flags, self.stages):
      matching.finalize()
      flagging.finalize()
      stage.finalize()
    for stack in (self.history, self.power, self.next_birth): stack.finalize()
    for key, value in (("pal.slot", self.slot), ("pal.small", self.small), ("pal.first", self.first)):
      self.circuit.put(key, value)
    self.circuit.put("pal.power_ready", Value.select(self.power_ready, Value.constant(True), Value.constant(False)))


def build(rates=DEFAULT_BATCH):
  circuit = Circuit("ab")
  controller = WindowPAL(circuit, rates)
  controller.tick()
  controller.finalize()
  return circuit, controller, circuit.machine(controller.output, initial_accepting=True)

File: scala/pal/src/main/scala/pal/ScaffoldWindowPal.scala
package pal

import Expr.{FALSE, TRUE}
import Ref.{EMPTY, PREVIOUS}
import ScaffoldCircuit.{choose, conjunction as AND, disjunction as OR, neg}

/** One reusable dyadic PAL stage and its four bounded result packets. */
final class WindowStage(val circuit: Circuit, val index: Int, historyPool: StackPool, val worker: WindowWorkerLike) {
  private val stageName = s"stage$index"
  val half: Stack = new Stack(historyPool, stageName + ".half", Some("cells"))
  val clock: Stack = new Stack(historyPool, stageName + ".clock", Some("cells"))
  val aliveKey: String = stageName + ".alive"
  val intervalKey: String = stageName + ".interval"
  var alive: Expr = circuit.get(PREVIOUS, aliveKey, Vector(false, true), false).eqTo(true)
  var interval: Value[Int] = circuit.get(PREVIOUS, intervalKey, (0 to 6).toVector, 0)
  val pendingKey: String = stageName + ".pending"
  val batchKey: String = stageName + ".batch"
  var pending: Expr = circuit.get(PREVIOUS, pendingKey, Vector(false, true), false).eqTo(true)
  var batch: Value[Int] = circuit.get(PREVIOUS, batchKey, (0 to 3).toVector, 0)
  val results: Vector[FlagStack] = Vector.tabulate(4)(i => new FlagStack(worker.flagPool.get, s"$stageName.result$i"))
  var birth: Expr = FALSE
  var release: Expr = FALSE
  var middle: Expr = FALSE

  def answering(): Expr = AND(alive, OR((2 to 5).map(interval.eqTo)*))

  def advance(newBirth: Expr, sourceHalf: Stack): Unit = {
    clock.drop(alive)
    val boundary = AND(alive, clock.empty())
    interval = Value.select(boundary, interval.map(n => math.min(n + 1, 6)), interval)
    clock.copyFrom(half, boundary)
    val newRelease = AND(boundary, OR((1 to 4).map(interval.eqTo)*))
    circuit.require(neg(pending), newRelease)
    batch = Value.select(newRelease, interval.map(n => math.max(0, math.min(3, n - 1))).recode((0 to 3).toVector), batch)
    alive = AND(alive, neg(AND(boundary, interval.eqTo(6))))
    half.copyFrom(sourceHalf, newBirth)
    clock.copyFrom(sourceHalf, newBirth)
    interval = Value.select(newBirth, Value.constant(0), interval)
    alive = OR(alive, newBirth)
    birth = newBirth
    release = AND(newRelease, neg(newBirth))
    pending = AND(pending, neg(newBirth))
    results.foreach(_.clear(newBirth))
  }

  def consume(): Unit = {
    val active = answering()
    val selected = new FlagStack(results.head.pool, results.head.name)
    selected.root = EMPTY
    selected.index = worker.flagPool.get.index(-1)
    for ((result, resultIndex) <- results.zipWithIndex) {
      selected.copyFrom(result, interval.eqTo(resultIndex + 2))
    }
    middle = AND(active, selected.pop(active))
    for ((result, resultIndex) <- results.zipWithIndex) {
      result.copyFrom(selected, AND(active, interval.eqTo(resultIndex + 2)))
    }
  }

  def capture(): Unit = {
    pending = OR(pending, release)
    val done = AND(pending, worker.mode.eqTo("done"))
    for ((result, resultIndex) <- results.zipWithIndex) {
      result.copyFrom(worker.flags.get, AND(done, batch.eqTo(resultIndex)))
    }
    pending = AND(pending, neg(done))
  }

  def commit(): Unit = {
    half.commit()
    clock.commit()
    results.foreach(_.commit())
    circuit.put(intervalKey, interval)
    circuit.put(batchKey, batch)
    circuit.put(aliveKey, Value.select(alive, Value.constant(true), Value.constant(false)))
    circuit.put(pendingKey, Value.select(pending, Value.constant(true), Value.constant(false)))
  }
}

/** Two dyadic PAL stages in one transition per original binary input. */
final class WindowPAL(val circuit: Circuit, val rates: BatchRates = GsBatchClock.DEFAULT_BATCH,
                      matcherFactory: (Circuit, String, Int) => WindowWorkerLike = ScaffoldWindowWorkers.matcher,
                      flagsFactory: (Circuit, String, Int) => WindowWorkerLike = ScaffoldWindowWorkers.flags) {
  private val defaults = GsBatchClock.DEFAULT_BATCH
  if (rates.k != defaults.k || rates.matching < defaults.matching || rates.flags < defaults.flags) {
    throw new IllegalArgumentException("the complete PAL source requires the derived k=8 service bounds")
  }
  val matchers: Vector[WindowWorkerLike] = Vector.tabulate(2)(i => matcherFactory(circuit, s"match$i", rates.matching))
  val flags: Vector[WindowWorkerLike] = Vector.tabulate(2)(i => flagsFactory(circuit, s"flag$i", rates.flags))
  private val historyPool = new StackPool(circuit, Vector("cells" -> 1))
  val history: Stack = new Stack(historyPool, "pal.history", Some("cells"))
  val power: Stack = new Stack(historyPool, "pal.power", Some("cells"))
  val nextBirth: Stack = new Stack(historyPool, "pal.next_birth", Some("cells"))
  val stages: Vector[WindowStage] = Vector.tabulate(2)(i => new WindowStage(circuit, i, historyPool, flags(i)))
  var powerReady: Expr = circuit.get(PREVIOUS, "pal.power_ready", Vector(false, true), false).eqTo(true)
  var slot: Value[Int] = circuit.get(PREVIOUS, "pal.slot", Vector(0, 1), 0)
  var small: Value[Int] = circuit.get(PREVIOUS, "pal.small", (0 to 4).toVector, 0)
  var first: Value[Boolean] = circuit.get(PREVIOUS, "pal.first", Vector(false, true), false)
  var output: Expr = TRUE

  def tick(): Unit = {
    val current = circuit.input().eqTo('b')
    first = Value.select(small.eqTo(0), Value.select(current, Value.constant(true), Value.constant(false)), first)
    small = small.map(n => math.min(4, n + 1))
    history.push()
    val firstRound = neg(powerReady)
    nextBirth.drop(powerReady)
    val birth = AND(powerReady, nextBirth.empty())
    for ((stage, stageIndex) <- stages.zipWithIndex) {
      stage.advance(AND(birth, slot.eqTo(stageIndex)), power)
    }
    power.copyFrom(history, OR(firstRound, birth))
    nextBirth.copyFrom(history, OR(firstRound, birth))
    powerReady = TRUE
    slot = Value.select(birth, slot.cycle(), slot)

    for (((stage, matching), flagging) <- stages.zip(matchers).zip(flags)) {
      matching.arrive()
      flagging.arrive()
      flagging.resetFlags(stage.birth)
      matching.start(stage.birth)
      flagging.start(stage.release)
      flagging.mark(stage.release)
      stage.consume()
      matching.service()
      flagging.service()
      stage.capture()
    }

    val active = stages.map(_.answering())
    circuit.require(AND(OR(active*), neg(AND(active*))), small.eqTo(4))
    val ordinary = OR(active.zip(stages).zip(matchers).map { case ((enabled, stage), matching) =>
      AND(enabled, stage.middle, matching.output)
    }*)
    val short = OR(small.eqTo(1), AND(OR(small.eqTo(2), small.eqTo(3)),
      choose(current, first.eqTo(true), first.eqTo(false))))
    output = choose(small.eqTo(4), ordinary, short)
  }

  def commit(): Unit = {
    for (((matching, flagging), stage) <- matchers.zip(flags).zip(stages)) {
      matching.commit()
      flagging.commit()
      stage.commit()
    }
    Vector(history, power, nextBirth).foreach(_.commit())
    circuit.put("pal.slot", slot)
    circuit.put("pal.small", small)
    circuit.put("pal.first", first)
    circuit.put("pal.power_ready", Value.select(powerReady, Value.constant(true), Value.constant(false)))
  }
}

object ScaffoldWindowPal {
  def build(rates: BatchRates = GsBatchClock.DEFAULT_BATCH): (Circuit, WindowPAL, Scaffold) = {
    val circuit = new Circuit("ab")
    val controller = new WindowPAL(circuit, rates)
    controller.tick()
    controller.commit()
    (circuit, controller, circuit.machine(controller.output, initialAccepting = true))
  }
}
